python/tutorials/02-fused-softmax.py

```diff
@@ -139,35 +139,32 @@ def softmax(x):
     y = torch.empty_like(x)
 
     # pre-compile kernel to get register usage and compute thread occupancy.
-    kernel, num_programs = kernels.get(BLOCK_SIZE, (None, 0))
-    if kernel is None:
-        kernel = softmax_kernel.warmup(y, x, x.stride(0), y.stride(0), n_rows, n_cols, BLOCK_SIZE=BLOCK_SIZE,
-                                       num_stages=num_stages, num_warps=num_warps, grid=(1, ))
-        kernel._init_handles()
-        n_regs = kernel.n_regs
-        size_smem = kernel.metadata.shared
-        if is_hip():
-            # NUM_REGS represents the number of regular purpose registers. On CDNA architectures this is half of all registers available.
-            # However, this is not always the case. In most cases all registers can be used as regular purpose registers.
-            # ISA SECTION (3.6.4 for CDNA3)
-            # VGPRs are allocated out of two pools: regular VGPRs and accumulation VGPRs. Accumulation VGPRs are used
-            # with matrix VALU instructions, and can also be loaded directly from memory. A wave may have up to 512 total
-            # VGPRs, 256 of each type. When a wave has fewer than 512 total VGPRs, the number of each type is flexible - it is
-            # not required to be equal numbers of both types.
-            if is_cdna():
-                NUM_GPRS = NUM_REGS * 2
-
-            # MAX_NUM_THREADS represents maximum number of resident threads per multi-processor.
-            # When we divide this number with WARP_SIZE we get maximum number of waves that can
-            # execute on a CU (multi-processor)  in parallel.
-            MAX_NUM_THREADS = properties["max_threads_per_sm"]
-            max_num_waves = MAX_NUM_THREADS // WARP_SIZE
-            occupancy = min(NUM_GPRS // WARP_SIZE // n_regs, max_num_waves) // num_warps
-        else:
-            occupancy = NUM_REGS // (n_regs * WARP_SIZE * num_warps)
-        occupancy = min(occupancy, SIZE_SMEM // size_smem)
-        num_programs = NUM_SM * occupancy
-        kernels[BLOCK_SIZE] = (kernel, num_programs)
+    kernel = softmax_kernel.warmup(y, x, x.stride(0), y.stride(0), n_rows, n_cols, BLOCK_SIZE=BLOCK_SIZE,
```

Comment:
I think there is still a bug here because warmup uses `MockTensor`:
https://github.com/triton-lang/triton/blob/1cf06c5e1982eba8f17062e1c6c3d3fa458597b2/python/triton/runtime/jit.py#L729

And `MockTensor` doesn't respect the real pointer's alignment:
https://github.com/triton-lang/triton/blob/1cf06c5e1982eba8f17062e1c6c3d3fa458597b2/python/triton/runtime/jit.py#L879-L881

This seems questionable though. I'm not sure why warmup couldn't just operate on the real tensors since it doesn't actually run any device code.

Reply:
Nice catch! Let me try digging into this

Reply:
So I dug into this more. Since `kernel.warmup` is being called with the actual tensors, they don't get mapped to MockTensor. Arguments are only mapped to MockTensor when a dtype is passed, e.g. `kernel.warmup(torch.float32, torch.int32)`.

MockTensor making an optimistic assumption about the pointer alignment is also OK. It just means if the kernel is called with a pointer with a different alignment, the JIT will cache miss and compile the kernel again anyways.

WDYT about removing the caching mechanism in the tutorial, however?

Reply:
Ah my mistake, sorry for the noise.